aptos-move/framework/src/natives/signature.rs

```diff
@@ -32,22 +149,65 @@ pub fn native_bls12381_public_key_validation(
     let pop_bytes = pop_arg!(arguments, Vec<u8>);
     let key_bytes = pop_arg!(arguments, Vec<u8>);
 
+    // NOTE(Gas): O(|pop_bytes|) deserialization cost
     let pop = match bls12381::ProofOfPossession::try_from(&pop_bytes[..]) {
         Ok(pop) => pop,
         Err(_) => return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
     };
 
+    // NOTE(Gas): O(|key_bytes|) deserialization cost
     let public_key = match bls12381::PublicKey::try_from(&key_bytes[..]) {
         Ok(key) => key,
         Err(_) => return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
     };
 
+    // NOTE(Gas): O(1) cost: 2 bilinear pairings and a hash-to-curve
     let valid = pop.verify(&public_key).is_ok();
 
     Ok(NativeResult::ok(cost, smallvec![Value::bool(valid)]))
 }
 
-pub fn native_ed25519_publickey_validation(
+pub fn native_bls12381_verify_signature(
+    _context: &mut NativeContext,
+    _ty_args: Vec<Type>,
+    mut arguments: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(_ty_args.is_empty());
+    debug_assert!(arguments.len() == 3);
+
+    // TODO: replace with proper gas cost
+    let cost = GasCost::new(super::cost::APTOS_LIB_TYPE_OF, 1).total();
+
+    let msg_bytes = pop_arg!(arguments, Vec<u8>);
+    let pk_bytes = pop_arg!(arguments, Vec<u8>);
+    let sig_bytes = pop_arg!(arguments, Vec<u8>);
+
+    // NOTE(Gas): O(|pk_bytes|) deserialization cost
```

Comment:
Not necessarily related to this PR, but do these public keys have constant sizes? If so, I think we have get away with having only a base_cost for gas metering.

Reply:
Yes, they do have constant-size. Should I update all these costs to O(1) in that case?

Reply:
Fixed.